src/include/duckdb/execution/operator/join/physical_hash_join.hpp

```diff
@@ -9,22 +9,60 @@
 #pragma once
 
 #include "duckdb/common/types/chunk_collection.hpp"
+#include "duckdb/common/value_operations/value_operations.hpp"
 #include "duckdb/execution/join_hashtable.hpp"
+#include "duckdb/execution/operator/join/perfect_hash_join_executor.hpp"
 #include "duckdb/execution/operator/join/physical_comparison_join.hpp"
 #include "duckdb/execution/physical_operator.hpp"
 #include "duckdb/planner/operator/logical_join.hpp"
 
 namespace duckdb {
 
+class PhysicalHashJoinState : public PhysicalOperatorState {
+public:
+	PhysicalHashJoinState(PhysicalOperator &op, PhysicalOperator *left, PhysicalOperator *right,
+	                      vector<JoinCondition> &conditions)
+	    : PhysicalOperatorState(op, left) {
+	}
+
+	DataChunk cached_chunk;
+	DataChunk join_keys;
+	ExpressionExecutor probe_executor;
+	unique_ptr<JoinHashTable::ScanStructure> scan_structure;
+	SelectionVector build_sel_vec;
+	SelectionVector probe_sel_vec;
+	SelectionVector seq_sel_vec;
+};
 //! PhysicalHashJoin represents a hash loop join between two tables
+
+class HashJoinGlobalState : public GlobalOperatorState {
+public:
+	HashJoinGlobalState() {
+	}
+
+	//! The HT used by the join
+	unique_ptr<JoinHashTable> hash_table;
+	//! Only used for FULL OUTER JOIN: scan state of the final scan to find unmatched tuples in the build-side
+	JoinHTScanState ht_scan_state;
+	vector<Vector> build_columns; // for build_data storage
+	LogicalType key_type {LogicalType::INVALID};
```

Comment:
Any reason this is in the global state? This should be constant and known before-hand, no?

Reply:
This wasn't supposed to be there. My bad.